core/src/forest/ForestTrainer.cpp

```diff
@@ -135,8 +134,9 @@ std::vector<std::shared_ptr<Tree>> ForestTrainer::train_ci_group(Data* data,
                                                                  const ForestOptions& options) const {
   std::vector<std::shared_ptr<Tree>> trees;
 
+  // If clustering enabled, samples by cluster id rather than by observation id
   std::vector<size_t> sample;
-  sampler.sample(data->get_num_rows(), 0.5, sample);
+  sampler.sample_for_ci(data, 0.5, sample);
```

Comment:
We want to always use the same clustering strategy, whether `ci_group_size` is >1 or not. Sampling by groups changes the estimands (i.e., do we give equal weights to clusters or equal weights to samples), so this matters even if we're not building CIs.